network/p2p/inspector/validation/control_message_validation_inspector.go

```diff
@@ -209,6 +209,24 @@ func (c *ControlMsgValidationInspector) Inspect(from peer.ID, rpc *pubsub.RPC) e
 		return nil
 	}
 
+	// check peer identity when running private network
+	// sanity check: rpc inspection should be disabled on public networks
+	if c.networkingType == network.PrivateNetwork && c.config.InspectionProcess.Inspect.RejectUnstakedPeers {
+		err := c.checkSenderIdentity(from)
```

Comment:
how often is this check? do we need to cache it? If a unstaked peer got rejected, and try to connect again, would the peer get the same pid again?

Reply:
We don't need to cache it here because if a sender is unstaked it will be pruned on the next gossipsub [heart beat](https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L1403). During each heart beat gossipsub will query the gossip score for all peers in the nodes mesh from our [scoring registry](https://github.com/onflow/flow-go/blob/master/network/p2p/scoring/registry.go#L261), we [apply the maximum](https://github.com/onflow/flow-go/blob/master/network/p2p/scoring/registry.go#L351) penalty to a nodes score for unknown identities so, so we will reject RPC's for 1 or 2 heartbeats. The heartbeat interval is [1 second](https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L44).